File: android/src/main/kotlin/com/abc/klc_klaviyo_flutter/KlcKlaviyoFlutterPlugin.kt
package com.abc.klc_klaviyo_flutter

import android.app.Activity
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.core.app.NotificationManagerCompat
import com.google.firebase.messaging.FirebaseMessaging
import com.klaviyo.analytics.Klaviyo
import com.klaviyo.analytics.model.Event
import com.klaviyo.analytics.model.EventMetric
import com.klaviyo.forms.registerForInAppForms
import com.klaviyo.forms.unregisterFromInAppForms
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class KlcKlaviyoFlutterPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
    
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private var activity: Activity? = null

    companion object {
        private const val TAG = "KlaviyoFlutterPlugin"
        private const val CHANNEL_NAME = "com.klaviyo.flutter/klaviyo_sdk"
        private var instance: KlcKlaviyoFlutterPlugin? = null
        
        // Để KlaviyoPushService gọi được
        fun sendPushTokenToFlutter(token: String) {
            instance?.let { plugin ->
                Handler(Looper.getMainLooper()).post {
                    plugin.channel.invokeMethod("onPushTokenReceived", mapOf("token" to token))
                }
            }
        }
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        Log.d(TAG, "🔌 Plugin attached to engine")
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
        context = binding.applicationContext
        instance = this
        Log.d(TAG, "✅ Plugin initialized successfully")
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        instance = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        Log.d(TAG, "📞 Method called: ${call.method}")
        when (call.method) {
            "initialize" -> handleInitialize(call, result)
            "setEmail" -> handleSetEmail(call, result)
            "setPhoneNumber" -> handleSetPhoneNumber(call, result)
            "setExternalId" -> handleSetExternalId(call, result)
            "setProfile" -> handleSetProfile(call, result)
            "resetProfile" -> handleResetProfile(result)
            "createEvent" -> handleCreateEvent(call, result)
            "setPushToken" -> handleSetPushToken(call, result)
            "getPushToken" -> handleGetPushToken(result)
            "getEmail" -> result.success(Klaviyo.getEmail())
            "getPhoneNumber" -> result.success(Klaviyo.getPhoneNumber())
            "getExternalId" -> result.success(Klaviyo.getExternalId())
            "requestPushPermission" -> handleRequestPushPermission(result)
            "getPushPermissionStatus" -> handleGetPushPermissionStatus(result)
            "registerForInAppForms" -> handleRegisterForInAppForms(result)
            "unregisterFromInAppForms" -> handleUnregisterFromInAppForms(result)
            else -> {
                Log.w(TAG, "⚠️ Method not implemented: ${call.method}")
                result.notImplemented()
            }
        }
    }

    private fun handleInitialize(call: MethodCall, result: Result) {
        Log.d(TAG, "🚀 Initializing Klaviyo...")
        val apiKey = call.argument<String>("apiKey")
        if (apiKey.isNullOrEmpty()) {
            result.error("INVALID_ARGS", "API key required", null)
            return
        }
        try {
            Klaviyo.initialize(apiKey, context)
            Log.d(TAG, "✅ Klaviyo initialized with API key: ${apiKey.take(6)}...")
            result.success(null)
        } catch (e: Exception) {
            result.error("INIT_ERROR", e.message, null)
        }
    }

    private fun handleSetEmail(call: MethodCall, result: Result) {
        val email = call.argument<String>("email")
        if (email.isNullOrEmpty()) {
            result.error("INVALID_ARGS", "Email required", null)
            return
        }
        Klaviyo.setEmail(email)
        result.success(null)
    }

    private fun handleSetPhoneNumber(call: MethodCall, result: Result) {
        val phone = call.argument<String>("phoneNumber")
        if (phone.isNullOrEmpty()) {
            result.error("INVALID_ARGS", "Phone required", null)
            return
        }
        Klaviyo.setPhoneNumber(phone)
        result.success(null)
    }

    private fun handleSetExternalId(call: MethodCall, result: Result) {
        val id = call.argument<String>("externalId")
        if (id.isNullOrEmpty()) {
            result.error("INVALID_ARGS", "External ID required", null)
            return
        }
        Klaviyo.setExternalId(id)
        result.success(null)
    }

    private fun handleSetProfile(call: MethodCall, result: Result) {
        Log.d(TAG, "👤 Setting profile...")
        try {
            call.argument<String>("email")?.let {
                Log.d(TAG, "  ✉️ Email: $it")
                Klaviyo.setEmail(it)
            }
            call.argument<String>("phoneNumber")?.let {
                Log.d(TAG, "  📱 Phone: $it")
                Klaviyo.setPhoneNumber(it)
            }
            call.argument<String>("externalId")?.let {
                Log.d(TAG, "  🆔 External ID: $it")
                Klaviyo.setExternalId(it)
            }
            Log.d(TAG, "✅ Profile set successfully")
            result.success(null)
        } catch (e: Exception) {
            result.error("SET_PROFILE_ERROR", e.message, null)
        }
    }

    private fun handleResetProfile(result: Result) {
        Klaviyo.resetProfile()
        result.success(null)
    }

    private fun handleCreateEvent(call: MethodCall, result: Result) {
        val name = call.argument<String>("name")
        Log.d(TAG, "📊 Creating event: $name")
        if (name.isNullOrEmpty()) {
            Log.e(TAG, "❌ Event name is empty")
            result.error("INVALID_ARGS", "Event name required", null)
            return
        }

        try {
            val event = Event(EventMetric.CUSTOM(name))
            Klaviyo.createEvent(event)
            Log.d(TAG, "✅ Event created: $name")
            result.success(null)
        } catch (e: Exception) {
            result.error("CREATE_EVENT_ERROR", e.message, null)
        }
    }

    private fun handleSetPushToken(call: MethodCall, result: Result) {
        val token = call.argument<String>("token")
        if (token.isNullOrEmpty()) {
            result.error("INVALID_ARGS", "Token required", null)
            return
        }
        Klaviyo.setPushToken(token)
        result.success(null)
    }

    private fun handleGetPushToken(result: Result) {
        FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
            if (task.isSuccessful) {
                result.success(task.result)
            } else {
                result.success(null)
            }
        }
    }

    private fun handleRequestPushPermission(result: Result) {
        Log.d(TAG, "🔔 Requesting push permission...")
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            activity?.let { act ->
                Log.d(TAG, "  Requesting POST_NOTIFICATIONS permission")
                act.requestPermissions(
                    arrayOf(android.Manifest.permission.POST_NOTIFICATIONS),
                    1001
                )
            }
        }

        FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
            if (task.isSuccessful) {
                val token = task.result
                Klaviyo.setPushToken(token)
                channel.invokeMethod("onPushTokenReceived", mapOf("token" to token))
            } else {
                Log.e(TAG, "❌ Failed to get FCM token: ${task.exception?.message}")
            }
        }

        result.success(null)
    }

    private fun handleGetPushPermissionStatus(result: Result) {
        val enabled = NotificationManagerCompat.from(context).areNotificationsEnabled()
        result.success(if (enabled) "authorized" else "denied")
    }

    // MARK: - In-App Forms

    /// Registers for in-app forms to display when conditions are met
    private fun handleRegisterForInAppForms(result: Result) {
        Log.d(TAG, "📝 Registering for in-app forms...")
        try {
            Handler(Looper.getMainLooper()).post {
                Klaviyo.registerForInAppForms()
            }
            result.success(null)
        } catch (e: Exception) {
            Log.e(TAG, "❌ Register for in-app forms error: ${e.message}")
            result.error("REGISTER_FORMS_ERROR", e.message, null)
        }
    }

    /// Unregisters from in-app forms to stop displaying them
    private fun handleUnregisterFromInAppForms(result: Result) {
        Log.d(TAG, "📝 Unregistering from in-app forms...")
        try {
            Klaviyo.unregisterFromInAppForms()
            result.success(null)
        } catch (e: Exception) {
            Log.e(TAG, "❌ Unregister from in-app forms error: ${e.message}")
            result.error("UNREGISTER_FORMS_ERROR", e.message, null)
        }
    }
}

File: ios/Classes/KlcKlaviyoFlutterPlugin.swift
import Flutter
import UIKit
import KlaviyoSwift
import KlaviyoForms
import UserNotifications

public class KlcKlaviyoFlutterPlugin: NSObject, FlutterPlugin, UNUserNotificationCenterDelegate {

    private var channel: FlutterMethodChannel?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.klaviyo.flutter/klaviyo_sdk",
            binaryMessenger: registrar.messenger()
        )
        let instance = KlcKlaviyoFlutterPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)

        // Set as notification delegate to handle foreground notifications and taps
        UNUserNotificationCenter.current().delegate = instance
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handle: Method called - \(call.method)")
        switch call.method {
        case "initialize":
            handleInitialize(call, result: result)
        case "setEmail":
            handleSetEmail(call, result: result)
        case "setPhoneNumber":
            handleSetPhoneNumber(call, result: result)
        case "setExternalId":
            handleSetExternalId(call, result: result)
        case "setProfile":
            handleSetProfile(call, result: result)
        case "resetProfile":
            handleResetProfile(result: result)
        case "createEvent":
            handleCreateEvent(call, result: result)
        case "setPushToken":
            handleSetPushToken(call, result: result)
        case "getPushToken":
            handleGetPushToken(result: result)
        case "requestPushPermission":
            handleRequestPushPermission(result: result)
        case "getPushPermissionStatus":
            handleGetPushPermissionStatus(result: result)
        case "getEmail":
            result(KlaviyoSDK().email)
        case "getPhoneNumber":
            result(KlaviyoSDK().phoneNumber)
        case "getExternalId":
            result(KlaviyoSDK().externalId)
        case "registerForInAppForms":
            handleRegisterForInAppForms(result: result)
        case "unregisterFromInAppForms":
            handleUnregisterFromInAppForms(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - Initialize
    
    private func handleInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handleInitialize: Starting initialization")
        guard let args = call.arguments as? [String: Any],
              let apiKey = args["apiKey"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "API key is required", details: nil))
            return
        }
        KlaviyoSDK().initialize(with: apiKey)
        print("[KlcKlaviyoFlutterPlugin] handleInitialize: SDK initialized successfully")
        result(nil)
    }
    
    // MARK: - Profile Management
    
    private func handleSetEmail(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handleSetEmail: Setting email")
        guard let args = call.arguments as? [String: Any],
              let email = args["email"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Email is required", details: nil))
            return
        }
        KlaviyoSDK().set(email: email)
        reRegisterStoredPushTokenIfAvailable()
        print("[KlcKlaviyoFlutterPlugin] handleSetEmail: Email set successfully")
        result(nil)
    }
    
    private func handleSetPhoneNumber(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handleSetPhoneNumber: Setting phone number")
        guard let args = call.arguments as? [String: Any],
              let phoneNumber = args["phoneNumber"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Phone number is required", details: nil))
            return
        }
        KlaviyoSDK().set(phoneNumber: phoneNumber)
        reRegisterStoredPushTokenIfAvailable()
        print("[KlcKlaviyoFlutterPlugin] handleSetPhoneNumber: Phone number set successfully")
        result(nil)
    }
    
    private func handleSetExternalId(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handleSetExternalId: Setting external ID")
        guard let args = call.arguments as? [String: Any],
              let externalId = args["externalId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "External ID is required", details: nil))
            return
        }
        KlaviyoSDK().set(externalId: externalId)
        reRegisterStoredPushTokenIfAvailable()
        print("[KlcKlaviyoFlutterPlugin] handleSetExternalId: External ID set successfully")
        result(nil)
    }
    
    private func handleSetProfile(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handleSetProfile: Setting profile")
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Profile data is required", details: nil))
            return
        }

        var location: Profile.Location?
        if let locationData = args["location"] as? [String: Any] {
            location = Profile.Location(
                address1: locationData["address1"] as? String,
                address2: locationData["address2"] as? String,
                city: locationData["city"] as? String,
                country: locationData["country"] as? String,
                latitude: locationData["latitude"] as? Double,
                longitude: locationData["longitude"] as? Double,
                region: locationData["region"] as? String,
                zip: locationData["zip"] as? String,
                timezone: locationData["timezone"] as? String
            )
        }
        
        let profile = Profile(
            email: args["email"] as? String,
            phoneNumber: args["phoneNumber"] as? String,
            externalId: args["externalId"] as? String,
            firstName: args["firstName"] as? String,
            lastName: args["lastName"] as? String,
            organization: args["organization"] as? String,
            title: args["title"] as? String,
            image: args["image"] as? String,
            location: location,
            properties: args["properties"] as? [String: Any]
        )
        
        KlaviyoSDK().set(profile: profile)
        reRegisterStoredPushTokenIfAvailable()
        print("[KlcKlaviyoFlutterPlugin] handleSetProfile: Profile set successfully")
        result(nil)
    }
    
    private func handleResetProfile(result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handleResetProfile: Resetting profile")
        KlaviyoSDK().resetProfile()
        print("[KlcKlaviyoFlutterPlugin] handleResetProfile: Profile reset successfully")
        result(nil)
    }
    
    // MARK: - Event Tracking
    
    private func handleCreateEvent(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handleCreateEvent: Creating event")
        guard let args = call.arguments as? [String: Any],
              let eventName = args["name"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Event name is required", details: nil))
            return
        }

        let event = Event(
            name: .customEvent(eventName),
            properties: args["properties"] as? [String: Any],
            value: args["value"] as? Double,
            uniqueId: args["uniqueId"] as? String
        )

        KlaviyoSDK().create(event: event)
        print("[KlcKlaviyoFlutterPlugin] handleCreateEvent: Event created successfully")
        result(nil)
    }
    
    // MARK: - Push Notifications
    
    private func handleRequestPushPermission(result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handleRequestPushPermission: Requesting push permission")
        #if targetEnvironment(simulator)
        result(FlutterError(code: "SIMULATOR", message: "Push unavailable on Simulator", details: nil))
        return
        #else
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound, .badge]) { granted, error in
            if let error = error {
                result(FlutterError(code: "PERMISSION_ERROR", message: error.localizedDescription, details: nil))
                return
            }
            guard granted else {
                result(FlutterError(code: "PERMISSION_DENIED", message: "User denied permission", details: nil))
                return
            }
            DispatchQueue.main.async {
                UIApplication.shared.registerForRemoteNotifications()
                result(nil)
            }
        }
        #endif
    }
    
    private func handleGetPushPermissionStatus(result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handleGetPushPermissionStatus: Getting push permission status")
        #if targetEnvironment(simulator)
        result("unavailable_simulator")
        #else
        UNUserNotificationCenter.current().getNotificationSettings { settings in
            let status: String
            switch settings.authorizationStatus {
            case .notDetermined: status = "not_determined"
            case .denied: status = "denied"
            case .authorized: status = "authorized"
            case .provisional: status = "provisional"
            case .ephemeral: status = "ephemeral"
            @unknown default: status = "unknown"
            }
            result(status)
        }
        #endif
    }
    
    private func handleSetPushToken(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handleSetPushToken: Setting push token")
        guard let args = call.arguments as? [String: Any],
              let tokenString = args["token"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Push token is required", details: nil))
            return
        }
        let trimmed = tokenString.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            result(FlutterError(code: "INVALID_TOKEN", message: "Invalid push token", details: nil))
            return
        }
        KlaviyoSDK().set(pushToken: trimmed)
        UserDefaults.standard.set(trimmed, forKey: "klaviyo_push_token")
        print("[KlcKlaviyoFlutterPlugin] handleSetPushToken: Push token set successfully")
        result(nil)
    }
    
    private func reRegisterStoredPushTokenIfAvailable() {
        guard let token = UserDefaults.standard.string(forKey: "klaviyo_push_token"),
              !token.isEmpty else {
            return
        }
        KlaviyoSDK().set(pushToken: token)
        print("[KlcKlaviyoFlutterPlugin] reRegisterStoredPushTokenIfAvailable: Token re-registered successfully")
    }
    
    private func handleGetPushToken(result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handleGetPushToken: Getting stored push token")
        let token = UserDefaults.standard.string(forKey: "klaviyo_push_token")
        result(token)
    }
    
    // MARK: - In-App Forms
    
    private func handleRegisterForInAppForms(result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handleRegisterForInAppForms: Registering for in-app forms")
        Task { @MainActor in
            KlaviyoSDK().registerForInAppForms()
            result(nil)
        }
    }
    
    private func handleUnregisterFromInAppForms(result: @escaping FlutterResult) {
        print("[KlcKlaviyoFlutterPlugin] handleUnregisterFromInAppForms: Unregistering from in-app forms")
        Task { @MainActor in
            KlaviyoSDK().unregisterFromInAppForms()
            result(nil)
        }
    }
    
    // MARK: - Application Delegate (for push token)
    
    public func application(_ application: UIApplication,
                           didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
        print("[KlcKlaviyoFlutterPlugin] didRegisterForRemoteNotificationsWithDeviceToken: Received device token")
        let tokenString = deviceToken.map { String(format: "%02.2hhx", $0) }.joined()
        KlaviyoSDK().set(pushToken: deviceToken)
        UserDefaults.standard.set(tokenString, forKey: "klaviyo_push_token")
        channel?.invokeMethod("onPushTokenReceived", arguments: ["token": tokenString])
        print("[KlcKlaviyoFlutterPlugin] didRegisterForRemoteNotificationsWithDeviceToken: Token registered and sent to Flutter")
    }
    
    public func application(_ application: UIApplication,
                           didFailToRegisterForRemoteNotificationsWithError error: Error) {
        channel?.invokeMethod("onPushTokenRegistrationFailed", arguments: ["error": error.localizedDescription])
        print("[KlcKlaviyoFlutterPlugin] didFailToRegisterForRemoteNotificationsWithError: Error sent to Flutter")
    }

    // MARK: - UNUserNotificationCenterDelegate

    /// Handle notification when app is in FOREGROUND
    public func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        willPresent notification: UNNotification,
        withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void
    ) {
        let userInfo = notification.request.content.userInfo
        print("[KlcKlaviyoFlutterPlugin] userNotificationCenter:willPresent: Notification received (foreground): \(userInfo)")

        // Show notification banner even when app is in foreground
        if #available(iOS 14.0, *) {
            completionHandler([.banner, .sound, .badge])
        } else {
            completionHandler([.alert, .sound, .badge])
        }
    }

    /// Handle notification when user TAPS on it
    public func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        didReceive response: UNNotificationResponse,
        withCompletionHandler completionHandler: @escaping () -> Void
    ) {
        let userInfo = response.notification.request.content.userInfo

        // Handle Klaviyo push opened event
        KlaviyoSDK().handle(notificationResponse: response, withCompletionHandler: completionHandler)

        // Notify Flutter about the tap
        channel?.invokeMethod("onNotificationTapped", arguments: userInfo)
        print("[KlcKlaviyoFlutterPlugin] userNotificationCenter:didReceive: Notification tap sent to Flutter")
    }
}
